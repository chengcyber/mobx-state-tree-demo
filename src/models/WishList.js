import { types, getParent, destroy } from 'mobx-state-tree'

const data = {
    name: 'Chronsiles of Noslk',
    price: 28.73,
    image: 'http://lorempixel.com/640/480/?15509',
}

export const WishListItem = types.model({
    name: types.string,
    price: types.number,
    image: types.optional(types.string, ''),
}).actions(self => ({
    setName(n) { self.name = n },
    setPrice(p) { self.price = p },
    setImage(i) { self.image = i},
    remove() {
        // delegate to list.remove
        // list.items.xxx so the depth is 2
        getParent(self, 2).remove(self)
    }
}))

export const WishList = types.model({
    items: types.optional(types.array(WishListItem), []),
}).actions(self => ({
    add(item) { self.items.push(item) },
    remove(item) {
        // self.items.splice(self.items.indexOf(item), 1)
        destroy(item)
    },
})).views(self => ({
    get totalPrice() {
        return self.items.reduce((sum, entry) => sum + entry.price, 0) * 1
    }
}))

export default WishList
