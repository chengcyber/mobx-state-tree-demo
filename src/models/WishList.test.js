import { getSnapshot, onSnapshot, onPatch } from 'mobx-state-tree';
import { WishListItem, WishList } from "./WishList";

it('can create a instance of a model', () => {
    const item = WishListItem.create({
        name: 'Chronsiles of Noslk',
        price: 28.73,
    })

    expect(item.price).toBe(28.73)
    expect(item.image).toBe('')
    item.setName('Niv')
    expect(item.name).toBe('Niv')
})

it('can create a wishlist', () => {
    const list = WishList.create({
        items: [
            {
                name: 'Chronsiles of Noslk',
                price: 28.73,
            }
        ],
    })

    expect(list.items.length).toBe(1)
    expect(list.items[0].price).toBe(28.73)
})

it('can add an item to wishlist', () => {
    const list = WishList.create()
    const states = []
    onSnapshot(list, snapshot => {
        states.push(snapshot)
    })

    list.add(WishListItem.create({
        name: 'Hola',
        price: 203.1,
    }))

    expect(list.items.length).toBe(1)
    expect(list.items[0].name).toBe('Hola')
    expect(list.items[0].price).toBe(203.1)

    expect(getSnapshot(list)).toEqual({
        items: [
            {
                name: 'Hola',
                price: 203.1,
                image: '',
            }
        ]
    })

    // expect(getSnapshot(list)).toMatchSnapshot()

    expect(states).toMatchSnapshot()
})

it('can add an item to wishlist - 2', () => {
    const list = WishList.create()
    const patchs = []
    onPatch(list, patch => {
        patchs.push(patch)
    })

    list.add(WishListItem.create({
        name: 'Hola',
        price: 203.1,
    }))

    list.items[0].setName('kkk')

    expect(patchs).toMatchSnapshot()
})

it('can calculate total price', () => {
    const list = WishList.create({
        items: [
            {
                name: 'a',
                price: 7.35,
            },
            {
                name: 'b',
                price: 123.2,
            },
        ]
    })

    expect(list.totalPrice).toBe(7.35 + 123.2)
})