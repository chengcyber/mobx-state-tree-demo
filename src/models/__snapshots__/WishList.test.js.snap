// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add an item to wishlist - 2 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "Hola",
      "price": 203.1,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "kkk",
  },
]
`;

exports[`can add an item to wishlist 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Hola",
        "price": 203.1,
      },
    ],
  },
]
`;
